flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkRewriteSubQueryRule.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.rex.RexSubQuery;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.fun.SqlCountAggFunction;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.immutables.value.Value;
+
+import java.util.Optional;
+
+/**
+ * Planner rule that rewrites scalar query in filter like: {@code select * from T1 where (select
+ * count(*) from T2) > 0} to {@code select * from T1 where exists (select * from T2)}, which could
+ * be converted to SEMI join by {@link FlinkSubQueryRemoveRule}.
+ *
+ * <p>Without this rule, the original query will be rewritten to a filter on a join on an aggregate
+ * by {@link SubQueryRemoveRule}. the full logical plan is
+ *
+ * <pre>{@code
+ * LogicalProject(a=[$0], b=[$1], c=[$2])
+ * +- LogicalJoin(condition=[$3], joinType=[semi])
+ *    :- LogicalTableScan(table=[[x, source: [TestTableSource(a, b, c)]]])
+ *    +- LogicalProject($f0=[IS NOT NULL($0)])
+ *       +- LogicalAggregate(group=[{}], m=[MIN($0)])
+ *          +- LogicalProject(i=[true])
+ *             +- LogicalTableScan(table=[[y, source: [TestTableSource(d, e, f)]]])
+ * }</pre>
+ */
+@Value.Enclosing
+public class FlinkRewriteSubQueryRule
+        extends RelRule<FlinkRewriteSubQueryRule.FlinkRewriteSubQueryRuleConfig> {
+
+    public static final FlinkRewriteSubQueryRule FILTER =
+            FlinkRewriteSubQueryRuleConfig.DEFAULT.toRule();
+
+    protected FlinkRewriteSubQueryRule(
+            FlinkRewriteSubQueryRule.FlinkRewriteSubQueryRuleConfig config) {
+        super(config);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        Filter filter = call.rel(0);
+        RexNode condition = filter.getCondition();
+        RexNode newCondition = rewriteScalarQuery(condition);
+        if (condition.equals(newCondition)) {
+            return;
+        }
+
+        Filter newFilter = filter.copy(filter.getTraitSet(), filter.getInput(), newCondition);
+        call.transformTo(newFilter);
+    }
+
+    // scalar query like: `(select count(*) from T) > 0` can be converted to `exists(select * from
+    // T)`
+    private RexNode rewriteScalarQuery(RexNode condition) {
+        return condition.accept(
+                new RexShuttle() {
+                    public RexNode visitCall(RexCall call) {
+                        Optional<RexSubQuery> subQuery = getSupportedScalarQuery(call);
+                        if (subQuery.isPresent()) {
+                            return RexSubQuery.exists(subQuery.get().rel.getInput(0));
+                        }
+                        return super.visitCall(call);
+                    }
+                });
+    }
+
+    private boolean isScalarQuery(RexNode n) {
+        return n.isA(SqlKind.SCALAR_QUERY);
+    }
+
```

Comment:
add this comment // check the RexNode is a RexLiteral which's value is between 0 and 1